### PR TITLE
Update Datetime & Ticket Sold Counts When Relations Change

### PR DESCRIPTION
**Pull request by tn3rb**
see: https://github.com/eventespresso/barista/issues/1088

This PR:

- adds a new method to EE_Base_Class for checking if a specific model object relation already exists

- updates sold and reserved counts for datetimes when ticket relations are added or removed

- removes previous logic for handling relation updates which only occurred when updating an event via the legacy event editor

- adds a unit test to confirm the above

### CODE REVIEW

File: core/db_classes/EE_Base_Class.class.php [hunk 1]
```diff
@@ -2250,6 +2250,27 @@ public function ID()
...
+    public function hasRelation($otherModelObjectOrID, string $relationName): bool
+    {
+        $other_model = self::_get_model_instance_with_name(
+            self::_get_model_classname($relationName),
+            $this->_timezone
+        );
+        $primary_key = $other_model->primary_key_name();
+        /** @var EE_Base_Class $otherModelObject */
+        $otherModelObject = $other_model->ensure_is_obj($otherModelObjectOrID, $relationName);
+        return $this->count_related($relationName, [[$primary_key => $otherModelObject->ID()]]) > 0;
```
Comment (manzoorwanijk) [inline core/db_classes/EE_Base_Class.class.php:2270]: 👍

File: core/db_classes/EE_Datetime.class.php [hunk 2]
```diff
@@ -1381,6 +1381,67 @@ public function venue(array $query_params = [])
...
+        $cache_id = null
+    ) {
+        // if we're adding a new relation to a ticket
+        if ($relationName === 'Ticket' && ! $this->hasRelation($otherObjectModelObjectOrID, $relationName)) {
+            /** @var EE_Ticket $ticket */
+            $ticket = EEM_Ticket::instance()->ensure_is_obj($otherObjectModelObjectOrID);
+            $this->increaseSold($ticket->sold(), false);
+            $this->increaseReserved($ticket->reserved());
+            $this->save();
+            $otherObjectModelObjectOrID = $ticket;
```
Comment (manzoorwanijk) [inline core/db_classes/EE_Datetime.class.php:1407]: Nice 👍

File: core/db_classes/EE_Datetime.class.php [hunk 3]
```diff
@@ -1381,6 +1381,67 @@ public function venue(array $query_params = [])
...
+     */
+    public function _remove_relation_to($otherObjectModelObjectOrID, $relationName, $where_query = [])
+    {
+        if ($relationName === 'Ticket' && $this->hasRelation($otherObjectModelObjectOrID, $relationName)) {
+            /** @var EE_Ticket $ticket */
+            $ticket = EEM_Ticket::instance()->ensure_is_obj($otherObjectModelObjectOrID);
+            $this->decreaseSold($ticket->sold());
+            $this->decreaseReserved($ticket->reserved());
+            $this->save();
+            $otherObjectModelObjectOrID = $ticket;
```
Comment (manzoorwanijk) [inline core/db_classes/EE_Datetime.class.php:1435]: Coolio

**Review by manzoorwanijk — APPROVED** [review]
Looks great 🚀

**Review by knazart — DISMISSED** [review]

**Comment by manzoorwanijk** [comment]
@tn3rb please check **Next steps** here https://github.com/eventespresso/barista/pull/1092

**Comment by tn3rb** [comment]
thnx Manzoor. The GQL mutations are fine as is, but I did need to add an override for `_remove_relations()` to both `EE_Datetime` and `EE_Ticket`, so that the sold and reserved counts also get updated when a related entity is completely removed.

File: core/db_classes/EE_Datetime.class.php [hunk 4]
```diff
@@ -1381,6 +1381,94 @@ public function venue(array $query_params = [])
...
+    {
+        if ($relationName === 'Ticket') {
+            $tickets = $this->tickets();
+            foreach ($tickets as $ticket) {
+                $this->decreaseSold($ticket->sold());
+                $this->decreaseReserved($ticket->reserved());
+                $this->save();
+            }
+        }
+        return parent::_remove_relations($relationName, $where_query_params);
```
Comment (manzoorwanijk) [inline core/db_classes/EE_Datetime.class.php:1468]: Cool. This should fix it for GQL mutations 👍

**Review by manzoorwanijk — APPROVED** [review]
Looks great now 🚀

**Pull request merged**